openssl/src/ssl/mod.rs

```diff
@@ -1223,6 +1227,44 @@ impl<S> MidHandshakeSslStream<S> {
         &self.error
     }
 
+    /// Gets the next DTLS handshake timeout.
+    ///
+    /// This is intended to be used with a nonblocking socket, where the value
+    /// can be used to set timeouts on reading and performing handshakes.
+    ///
+    /// `handle_dtlsv1_timeout` should be called when the timeout expires to
+    /// ask OpenSSL to retransmit.
+    ///
+    /// This method requires either the `dtlsv1` or `dtlsv1_2` feature.
+    #[cfg(any(feature = "dtlsv1", feature = "dtlsv1_2"))]
+    pub fn get_dtlsv1_timeout(&mut self) -> Option<time::Duration> {
+        let mut timeout: timeval = unsafe { mem::zeroed() };
+        let ret = unsafe { ffi::SSL_ctrl(self.stream.ssl.ssl, ffi::DTLS_CTRL_GET_TIMEOUT, 0,
+                                         &mut timeout as *mut _ as *mut _) };
+        if ret > 0 {
+            Some(time::Duration::new(timeout.tv_sec as u64, (timeout.tv_usec * 1000) as u32))
+        } else {
+            None
+        }
+    }
+
+    /// Handles DTLS handshake timeout expiry.
+    ///
+    /// This should be called when the timeout from `get_dtlsv1_timeout`
+    /// expires, to retransmit the handshake.
+    ///
+    /// This method requires either the `dtlsv1` or `dtlsv1_2` feature.
+    #[cfg(any(feature = "dtlsv1", feature = "dtlsv1_2"))]
+    pub fn handle_dtlsv1_timeout(&mut self) -> Result<(), Error> {
+        let ret = unsafe { ffi::SSL_ctrl(self.stream.ssl.ssl, ffi::DTLS_CTRL_HANDLE_TIMEOUT, 0,
+                                         ptr::null_mut()) };
+        if ret > 0 {
+            Ok(())
```

Comment:
Ick, I was reading off the boringssl headers for what the return values are: https://boringssl.googlesource.com/boringssl.git/+/master/include/openssl/ssl.h#523

Unfortunately, the documentation in the area is extremely sparse for actual OpenSSL, and that seems to be the most authoritative source on it. However, the OpenSSL library is pretty hard-and-fast on checking for > 0 rather than == 0 here, so I believe this is the correct behavior: https://github.com/openssl/openssl/blob/16938284cf5bba32006925281cb28b04d166a234/ssl/record/rec_layer_d1.c#L388